### DISCUSSION TITLE
ESP32 with stepper motor and driver TMC2209

### DISCUSSION THREAD

**Discussion by halfred** [micropython/micropython#11276] [ESP32]
Hello,

I'm trying to drive a [JK42HM ](https://datasheetspdf.com/pdf-file/1380129/ETC/JK42HM/1) stepper motor with a [TMC2209](https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC2209_Datasheet_V103.pdf) and an ESP32 programmed with micropython. Sometimes it works, sometimes it doesn't... and I can't figure out what is the difference between the 2 states (an important pin left unconnected?)

When it doesn't work, the motor either just shake or the 12V power supply emits a warning because a short circuit is detected. When it works, well.. it turns as expected

Here is my schema and the test code I use

```python
import time
import machine

dir = machine.Pin(2, machine.Pin.OUT)
p12 = machine.Pin(5, machine.Pin.OUT)
pwm12 = machine.PWM(p12)

dir.value(1)
pwm12.duty(512)

for i in range(100, 2000, 10):
    pwm12.freq(i)
    time.sleep(10)
``` 

![Schematic_TreeLamp_2023-04-12](https://user-images.githubusercontent.com/2283938/232277191-0f42f4a3-ea0c-4647-9668-ad6660c9a561.png)

I've searched a lot on the web but couldn't find something usefull :(
I'm probably not looking in the right direction.
Can anyone plesae help me ?

**Comment by robhamerling** [comment]
I have a similar setup and a similar problem!  At some specific PWM frequencies the motor doesn't step. My scope indicates a problem with the duty cycle. I set it once for 50%, but when changing the frequency the duty cycle changes sometimes (not always) to a strange value, even 0% or 100%. I'm trying to figure out when and why.

**Reply by robert-hh** [reply]
It it with ESP32 as well? And which firmware version?

**Comment by robhamerling** [comment]
Yes, it is with the ESP32 (one with 8 MB SPIRAM). Same script on the RP2 Pico does not show this behaviour.
The 'random' duty cycle appears even when explicitly setting it again to 50% after a frequency change. But it looks like a short wait after changing the frequency helps.

**Reply by robert-hh** [reply]
Is that just one period with a wrong cycle or a sequence?

**Reply by robert-hh** [reply]
An between which frequencies to you switch. I had seen non-matching pulse lengths when changing the frequency. The ESP32 does not seem to support smooth changes, like the RP2 does.

**Comment by robhamerling** [comment]
Once a 'wrong' duty cycle appears is remains the same during the cycle. When changing the frequency it usually switches back to 50%. 
I have a setup with a rotary to change the stepper frequency (in steps of about 106Hz):: 0-106-213-320-426, etc.  I see wrong duty cycles mostly at the frequencies 320, 640 and 1280 Hz
In the meantime I found out that a sleep of 1 PWM cycle between a frequency change and setting the duty cycle to 50% is enough to get rid of the duty cycle misbehaviour.

**Comment by robert-hh** [comment]
In my tests I have seen wrong duty cycles when the frequency change caused an internal binary divider to change. The 'critical' changes were for instance 610->611Hz, 1220->1221Hz, 4882->4883Hz. With your list of frequencies, 320, 640, 1280 are next to these limits, and changing the frequency down will cross them.  However, in my tests the change lasted only for a few periods, 1-2. So I wonder what causes the wrong cycle to be permanent and how Python code can change that behavior.
Below is an example for a 1220->1221 Hz change.

![pwm_1220_to_1221_Hz](https://github.com/micropython/micropython/assets/12476868/89e15adf-a9a8-4360-834b-a1480c396585)

**Comment by robhamerling** [comment]
In my original setup I made frequency steps of about 106 Hz. With steps of 1 Hz I didn't see irregularities, but with steps of 8 Hz the wrong duty cycles appeared again.  Until 152 Hz duty cycle was steady 50%.
From 152 to 160 Hz duty cycle became 100%!  then further to 168 Hz and somewhat higher it was 50% again
Backward from 168 to 160 Hz still 50%, then to 152 Hz: 25% (see picture).
I didn't try higher frequencies yet.

![SDS00009](https://github.com/micropython/micropython/assets/11457369/d90116ba-0f7c-413a-954a-6b7911faeda8)

**Reply by robert-hh** [reply]
Reading the data sheet, that should not happen if the PWM is set up properly. The ESP32 PWM module is under construction by @IhorNehrutsa, but that work stalled. So maybe this glitch will be fixed then.

**Comment by robhamerling** [comment]
I set the duty cycle with PWM.dutycycle_u16(32768) and by querying it after each frequency change (and a subsequent sleep of 1 PWM period) it reports most of the times 32678 as one would expect. But when I see on the scope a duty cycle of 100% it reports 65536 and when I see 25% it reports 16384. So it reports correctly the actual state!

**Reply by robert-hh** [reply]
So it looks like a code error. That may have been fixed in the most recent unmerged PR. Which MP version do you use?

**Comment by robhamerling** [comment]
Sorry, forgot to mention:
 MicroPython v1.20.0-151-ga61d40e5e-dirty on 2023-06-02; ESP32 module (spiram) with ESP32

**Reply by robert-hh** [reply]
So that's the most recent one. The best tip I can give you is what you probably do already: Set the duty rate again whenever you change the frequency,

**Comment by robert-hh** [comment]
I cannot replicate that using a GENERC_SPIRAM module and a fresh build. There is a short glitch upon the transition between 152 and 160 Hz (or 128 to 154) in both directions, but nothing permanent.

`MicroPython v1.20.0-151-ga61d40e5e on 2023-06-04; ESP32 module (spiram) with ESP32`

**Reply by robhamerling** [reply]
I'll to reduce my script to the bare minimum while it still shows the duty-cycle deviations.

**Reply by robhamerling** [reply]
I can reproduce the issue with 'varying' duty cycle in an REPL session.
I print a PWM instance (duty 50%) with freq=72, then change freq to 80, lastly freq back to 72 . 
From freq 72 to 80 duty cycle becomes 100%, back from 80 to 72 it becomes 25%.
```
>>> from machine import Pin, PWM
>>> pwm=PWM(Pin(5), freq=72, duty_u16=32768); print(pwm)
PWM(Pin(5), freq=72, duty_u16=32768, resolution=20, (duty=50.00%, resolution=0.000%), mode=0, channel=0, timer=0)
>>> pwm.freq(80); print(pwm)
PWM(Pin(5), freq=80, duty_u16=65536, resolution=19, (duty=100.00%, resolution=0.000%), mode=0, channel=0, timer=0)
>>> pwm.freq(72); print(pwm)
PWM(Pin(5), freq=72, duty_u16=16384, resolution=20, (duty=25.00%, resolution=0.000%), mode=0, channel=0, timer=0)
>>> 
```

**Reply by robert-hh** [reply]
I can reproduce it, but as well the "self-cure" after a short while:
```
>>> pwm.freq(72); print(pwm);time.sleep_ms(20);print(pwm)
PWM(Pin(5), freq=72, duty_u16=16384, resolution=20, (duty=25.00%, resolution=0.000%), mode=0, channel=0, timer=0)
PWM(Pin(5), freq=72, duty_u16=32768, resolution=20, (duty=50.00%, resolution=0.000%), mode=0, channel=0, timer=0)
>>> pwm.freq(80); print(pwm);time.sleep_ms(40);print(pwm)
PWM(Pin(5), freq=80, duty_u16=65536, resolution=19, (duty=100.00%, resolution=0.000%), mode=0, channel=0, timer=0)
PWM(Pin(5), freq=80, duty_u16=32768, resolution=19, (duty=50.00%, resolution=0.000%), mode=0, channel=0, timer=0)
```

**Reply by robhamerling** [reply]
Ah, forgot to switch on my scope....

**Comment by robhamerling** [comment]
I cannot reproduce the problem with a simple PWM setup. Preliminary conclusion: the cause of the issue with the varying duty_cycle must lie in my stepper-motor driver.

**Comment by halfred** [comment]
> I'm trying to drive a [JK42HM ](https://datasheetspdf.com/pdf-file/1380129/ETC/JK42HM/1) stepper motor with a [TMC2209](https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC2209_Datasheet_V103.pdf) and an ESP32 programmed with micropython. Sometimes it works, sometimes it doesn't... and I can't figure out what is the difference between the 2 states (an important pin left unconnected?)
> 
> When it doesn't work, the motor either just shake or the 12V power supply emits a warning because a short circuit is detected. When it works, well.. it turns as expected
> 
> Here is my schema and the test code I use
> 
> ```python
> import time
> import machine
> 
> dir = machine.Pin(2, machine.Pin.OUT)
> p12 = machine.Pin(5, machine.Pin.OUT)
> pwm12 = machine.PWM(p12)
> 
> dir.value(1)
> pwm12.duty(512)
> 
> for i in range(100, 2000, 10):
>     pwm12.freq(i)
>     time.sleep(10)
> ```
> 
> ![Schematic_TreeLamp_2023-04-12](https://user-images.githubusercontent.com/2283938/232277191-0f42f4a3-ea0c-4647-9668-ad6660c9a561.png)
> 
> I've searched a lot on the web but couldn't find something usefull :( I'm probably not looking in the right direction. Can anyone plesae help me ?

My initial issue is apparently solved.
I still have to make some tests, but it appears that the problems came from the motor driver.
I think that it was partially broken. That would explain the erratic behavior.

**Comment by robhamerling** [comment]
My stepper driver really needed removal of some proliferation! The stepper motor runs now without hick-ups.
I still don't understand the cause of the varying duty cycle, especially why it occurred at or around specific PWM frequencies, but not consistently. Probably a combination of improper pwm control and timings.

**Reply by OEngineer** [reply]
Is your schematic complete? The TMC2209 requires a couple of sense resistors as well as a number of capacitors for supply filtering and charge pump operation, according to the datasheet: 
![image](https://github.com/micropython/micropython/assets/49840280/50e70afb-0a25-444b-819b-9b4129540b96)

**Comment by Ika-Kvashali** [comment]
I was trying the same thing with TMC2209v1.2 and NEMA17 stepper motor, after increasing the speed of stepper motor the circuit worked for some time and after the driver got heated it stopped generating current, and now it does not even work on the lower speed. Is there any solution to this or anything that I am missing?
